src/store/level-wrapper.ts

```diff
@@ -170,6 +172,42 @@ export class LevelWrapper<V> {
     return abortOr(options?.signal, this.db.batch(operations));
   }
 
+  private async compactUnderlyingStorage(options?: LevelWrapperOptions): Promise<void> {
+    options?.signal?.throwIfAborted();
+
+    await abortOr(options?.signal, this.createLevelDatabase());
+
+    const range = this.sublevelRange;
+    if (!range) {
+      return;
+    }
+
+    // additional methods are only available on the root API instance
+    const root = this.root;
+
+    if (root.db.supports.additionalMethods.compactRange) {
+      return abortOr(options?.signal, root.db['compactRange']?.(...range));
+    }
+  }
+
+  private get sublevelRange(): [string, string] | undefined {
+    const prefix = this.db['prefix'] ?? '';
+    if (!prefix) {
+      return undefined;
+    }
+
+    // use the separator to derive an exclusive `end` that will never match to a key (which matches how `abstract-level` creates a `boundary`)
+    return [ prefix, prefix.slice(0, -1) + String.fromCharCode(prefix.charCodeAt(prefix.length - 1) + 1) ];
+  }
+
+  private get root(): LevelWrapper<V> {
+    let db = this.db;
+    for (const parent = db['db']; parent && parent !== db; ) {
```

Comment:
The `parent && parent !== db` throws me off... could `parent !== db` alone be enough?

Can you annotate in code why this check is the way it is? I can imagine why it "could" be needed, like:

- `db['db']` gets you the parent if there is a parent, which is kinda odd already, but fair enough I guess...
- when there is no parent, the `db['db']` can sometimes be `undefined` or sometimes be the same as the child db (this is the odd part for me)

without explicit comment, the code gives appearance of unnecessary check or potential bug.

Reply:
the interface of `AbstractSublevel` is where the `db` is, but some implementations of `AbstractLevel` will also provide `db` as a backwards compatibility thing

the `parent &&` is really just a defensive check since we dont want to try to create a `LevelWrapper` with a `db` that is `undefined`

Reply:
Repeating back what I understood, let me know if I got anything wrong:

- `db['db']` get you the level DB, it's the same instance of the level DB regardless of which "layer" of sublevel you are at
- at the root level (not sublevel), there is no guarantee that `db[`'db']` would return an instance (since it's an `AbstractSublevel` prop.

Reply:
i _think_ that `db['db']` will always get you the root `level`, but i'm not entirely sure so i wrote that code just in case `db['db']` only returns the parent.  e.g.
```
let db = new Level(...);
let a = db.sublevel('a');
let b = a.sublevel('b');
let c = b.sublevel('c');

console.log(c.db)  // should be `db`, but could be `b`
console.log(b.db)  // should be `db`, but could be `a`
console.log(a.db)  // should be `db`
console.log(db.db) // should be `undefined`, but could be `db` (see below)
```

according to `AbstractLevel` it shouldn't have a `db` property, but some implementationsprovide it for backwards compat reasons (e.g. [`browser-level`](https://github.com/Level/browser-level/blob/main/index.js#L67-L70))

TL;DR: i did this to be extra defensive against known scenarios where things aren't as described by the "types"